### ISSUE TITLE
XmlDocumentSource uses fixed character encoding (UTF8) for input XML streams. [CARROT-582]

### ISSUE THREAD

**Issue by dweiss** [carrot2/carrot2#787]
The `XmlDocumentSource` should respect regular &lt;?xml encoding="" ?&gt; headers. This will cause the DCS and command line applications to respect the encoding as well. For full correctness, we'd also need to add output encoding configuration entry to the DCS and command line applications. Ideally, the default output encoding would be the same as the encoding of the input XML, but I'm not sure if we can easily extract that (if so, it should become and output attribute of `XmlDocumentSource`).

---
Issue: CARROT-582 (migrated from JIRA), created by Dawid Weiss (@dweiss), resolved Dec 14 2009
Attachments: [carrot2-kaczynski.utf16.xml](https://carrotsearch.github.io/jira-archive-public/attachments/carrot2/CARROT-582/carrot2-kaczynski.utf16.xml), [carrot2-kaczynski.utf8.xml](https://carrotsearch.github.io/jira-archive-public/attachments/carrot2/CARROT-582/carrot2-kaczynski.utf8.xml), [CARROT-582.patch](https://carrotsearch.github.io/jira-archive-public/attachments/carrot2/CARROT-582/CARROT-582.patch), [simple-xml-1.7.2.jar](https://carrotsearch.github.io/jira-archive-public/attachments/carrot2/CARROT-582/simple-xml-1.7.2.jar)
Linked issues:
 - CARROT-583

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

Moving to 3.2.0.

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

Added a patch with JUnit test case + patched simple-xml-1.7.2 + test files.

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

The DCS should simply behave as a regular XML processing application – respect the encoding prologue. The output can be fixed to UTF-8 since it covers all Unicode and is size-efficient. We can safely assume applications that parse the output will respect the prologue as well (UTF-8 is the default encoding if it's missing), so I don't see the point of working on allowing custom output encodings.

**Comment by dweiss** [comment]
Comment by Stanisław Osiński (@stanislawosinski) (migrated from JIRA)

For XML there's no problem. The problem is for JSON. But that's a minor thing, nobody seems to be affected so far.

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

For JSON, it's an interesting thing because I have no idea how JavaScript evaluates a string retrieved using XMLHttpRequest. I assume (read: it really should be the case) the browser converts XmlHttpRequest's payload to unicode string (the only way to get hold of the payload) using HTTP header – content-type's charset declaration. If this is not the case, I'd say it's a bug in the browser.

In other words, for JSON, we should issue proper UTF-8 content-type and it should solve the problem, even if somebody has a different encoding on the AJAXish source page.

**Comment by dweiss** [comment]
Comment by Dawid Weiss (@dweiss) (migrated from JIRA)

In trunk.
